nibabel/cmdline/diff.py

```diff
@@ -0,0 +1,81 @@
+#!python
+# emacs: -*- mode: python-mode; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the NiBabel package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Output a summary table for neuroimaging files (resolution, dimensionality, etc.)
+"""
+from __future__ import division, print_function, absolute_import
+
+import sys
+from optparse import OptionParser, Option
+
+import numpy as np
+
+import nibabel as nib
+import nibabel.cmdline.utils
+from nibabel.cmdline.utils import _err, verbose, table2string, ap, safe_get
+import fileinput
+
+__author__ = 'Yaroslav Halchenko & Christopher Cheng'
+__copyright__ = 'Copyright (c) 2017 NiBabel contributors'
+__license__ = 'MIT'
+
+
+def get_opt_parser():
+    # use module docstring for help output
+    p = OptionParser(
+        usage="%s [OPTIONS] [FILE ...]\n\n" % sys.argv[0] + __doc__,
+        version="%prog " + nib.__version__)
+
+    p.add_options([
+        Option("-v", "--verbose", action="count",
+               dest="verbose", default=0,
+               help="Make more noise.  Could be specified multiple times"),
+
+        Option("-H", "--header-fields",
+               dest="header_fields", default='',
+               help="Header fields (comma separated) to be printed as well (if present)"),
+    ])
+
+    return p
+
+
+def diff_dicts(key, compare1, compare2):
+    """Returns the differences between two dicts"""
+    if np.any(compare1 != compare2):
+        return {key: (compare1, compare2)}
+    elif type(compare1) != type(compare2):
+        return {key: (compare1, compare2)}
+    else:
+        pass
+
+
+def main():
+    """Show must go on"""
+
+    parser = get_opt_parser()
+    (opts, files) = parser.parse_args()
+
+    nibabel.cmdline.utils.verbose_level = opts.verbose
+
+    if nibabel.cmdline.utils.verbose_level < 3:
+        # suppress nibabel format-compliance warnings
+        nib.imageglobals.logger.level = 50
+
+    assert len(files) == 2, "Please enter two files"
+    # TODO #3 -- make it work for any number
+
+    img1 = nib.load(files[0])
+    img2 = nib.load(files[1])
+
+    for i in img1.header.keys():
+        if diff_dicts(i, img1.header[i], img2.header[i]) is not None:
+            print(diff_dicts(i, img1.header[i], img2.header[i]))
```

Comment:
this looping should be within `diff_dicts` and I thought we had it. Moreover, keep in mind that there might be some keys present in one dict but not in the other... I thought we had those cases covered already

Reply:
Yeah those cases are covered to just return {'key': (x, None)}. At least I think it should. The list of header fields is accessed with code like img1.header['sizeof_hdr'], and if that field isn't available for either img1 or img2, I thought it should return a None. Please let me know if the truth is otherwise! And I'll work on popping the loop inside, thank you!

Reply:
Actually scratch that - I specifically left the loop out of the diff_values function because I thought the only way it could be tested in a separate file that tested specific inputs and outputs for diff_values would be to remove the accessing of the img.header fields from the program. I can bump this loop into another method but I'd otherwise be at a loss how to make both tests and the method work with the for loop inside. Could I please get your input on this?